### PR TITLE
Consultation features skeleton

### PR DESCRIPTION
**Pull request by wesho1107**
### Created several classes for consultation features:
Consultation, ConsultationList, ReadOnlyConsultationList, CreateConsultationCommand, CreateConsultationCreateParser

### Several changes to other parsers:
parseDate, parseTime, parseStudent, parseStudents, parseNames (several are not in used yet)

### Methods implemented:
**getMatchingNameStudent** and **matchName** to retrieve `Person` object matching the name input. Currently implemented to parse only one `Name`.
```
    @Override
    public Person getMatchingStudentName(Name name) {
        requireNonNull(name);
        return addressBook.matchName(name);
    }
```

### Commented code:
There are several portions of code yet to be implemented that are commented out.

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 1]
```diff
@@ -2,17 +2,17 @@
...
 import java.util.Set;
 
 import seedu.address.commons.core.index.Index;
 import seedu.address.commons.util.StringUtil;
 import seedu.address.logic.parser.exceptions.ParseException;
-import seedu.address.model.person.Address;
-import seedu.address.model.person.Email;
-import seedu.address.model.person.Name;
-import seedu.address.model.person.Phone;
+import seedu.address.model.person.*;
```
Comment (raydenlim) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: :( person.* 
import 1 by 1

File: src/main/java/seedu/address/model/consultation/Consultation.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
+package seedu.address.model.consultation;
+
+import seedu.address.model.person.Person;
+import seedu.address.model.person.UniquePersonList;
+
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.util.Set;
+
+public class Consultation {
```
Comment (raydenlim) [inline src/main/java/seedu/address/model/consultation/Consultation.java:14]: rmb to add javadocs for this file

File: src/main/java/seedu/address/model/consultation/ConsultationList.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
+package seedu.address.model.consultation;
```
Comment (raydenlim) [inline src/main/java/seedu/address/model/consultation/ConsultationList.java:1]: rmb to add javadocs

**Review by raydenlim — COMMENTED** [review]
LGTM,  just some minor stuff.
Rmb to add the javadocs into the files :)

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 1 above)
Comment (ldinghan) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java]: be careful when using intellij "smart" refactor

File: src/main/java/seedu/address/logic/Messages.java [hunk 4]
```diff
@@ -48,4 +49,17 @@ public static String format(Person person) {
...
 
+    /**
+     * Formats the {@code consultation} for display to the user.
+     */
+    public static String format(Consultation consultation) {
+        final StringBuilder builder = new StringBuilder();
+        builder.append(consultation.getDate())
+                .append("; Date: ")
+                .append(consultation.getTime())
+                .append("; Time: ");
```
Comment (ldinghan) [inline src/main/java/seedu/address/logic/Messages.java]: this seems to be in the wrong order, appears that you would put the date first, followed by its label e.g.

returns `2022-03-01; Date: 11:30; Time: ...`

File: src/main/java/seedu/address/logic/commands/consultation/CreateConsultCommand.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            + PREFIX_NAME + "Foo Bar";
+    public static final String MESSAGE_SUCCESS = "New consultation added: %1$s";
+    private final LocalDate date;
+    private final LocalTime time;
+    private Name name;
+    private Set<Name> names;
+    private Consultation consultationToAdd;
+
+    public CreateConsultCommand(LocalDate date, LocalTime time, Name name) {
+        requireNonNull(date);
```
Comment (ldinghan) [inline src/main/java/seedu/address/logic/commands/consultation/CreateConsultCommand.java]: can use `requireAllNonNull` instead to combine date, time, name requirements together

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 1 above)
Comment (Respirayson) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java]: you should update the settings else like intellij will group to * past a certain number of imports

File: src/main/java/seedu/address/model/ConsultationListBook.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        }
+
+        if (!(other instanceof ConsultationListBook)) {
+            return false;
```
Comment (ldinghan) [inline src/main/java/seedu/address/model/ConsultationListBook.java:90]: You can write a test case to cover this line by using ConsultationListBook.equals(*some stuff*)

File: src/main/java/seedu/address/model/consultation/Consultation.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        this.date = date;
+        this.time = time;
+        this.students = students;
+    }
+
+    //    /**
+    //     * Adds a student to the consultation.
+    //     * The person must not already exist in the consultation.
+    //     */
+    //    public void addStudent(Person student) {
```
Comment (ldinghan) [inline src/main/java/seedu/address/model/consultation/Consultation.java]: Not sure if this was left commented out intentionally?

File: src/main/java/seedu/address/model/consultation/ConsultationList.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        internalList.add(toAdd);
+    }
+
+    //    /**
+    //     * Replaces the consultation {@code target} in the list with {@code editedConsultation}.
+    //     * {@code target} must exist in the list.
+    //     * The consultation {@code editedConsultation} must not be the same as another existing
+    //     * consultation in the list.
+    //     */
+    //    public void setConsult(Consultation target, Consultation editedConsultation) {
```
Comment (ldinghan) [inline src/main/java/seedu/address/model/consultation/ConsultationList.java]: not sure if this was left commented out intentionally

File: src/main/java/seedu/address/model/consultation/ConsultationList.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) {
+            return true; // Same object, equal
+        }
+
+        if (!(other instanceof ConsultationList)) {
+            return false; // Not the same class, not equal
```
Comment (ldinghan) [inline src/main/java/seedu/address/model/consultation/ConsultationList.java:78]: same as for consultationlistbook, you can test this line by using ConsultationList.equals(*some other stuff*)

File: src/main/java/seedu/address/model/consultation/ConsultationList.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Override
+    public String toString() {
+        return internalList.toString();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) {
+            return true; // Same object, equal
```
Comment (ldinghan) [inline src/main/java/seedu/address/model/consultation/ConsultationList.java:74]: missed out this test case

**Review by ldinghan — COMMENTED** [review]
more test cases to add

File: src/main/java/seedu/address/model/consultation/ConsultationList.java (hunk 9 above)
Comment (wesho1107) [reply src/main/java/seedu/address/model/consultation/ConsultationList.java:78]: Resolved thanks!

File: src/main/java/seedu/address/model/consultation/ConsultationList.java (hunk 8 above)
Comment (wesho1107) [reply src/main/java/seedu/address/model/consultation/ConsultationList.java]: Removed deadcode for next milestone

File: src/main/java/seedu/address/model/ConsultationListBook.java (hunk 6 above)
Comment (wesho1107) [reply src/main/java/seedu/address/model/ConsultationListBook.java:90]: Covered!

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 11]
```diff
@@ -141,6 +144,28 @@ public static Set<Tag> parseTags(Collection<String> tags) throws ParseException
         return tagSet;
     }
 
+    /**
+     * Parses {@code String date} into an {@code LocalDate} and returns it. Leading and trailing whitespaces will be
+     * trimmed.
+     */
+    public static LocalDate parseDate(String date) throws ParseException {
```
Comment (Respirayson) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:154]: do you want to add in some exception checking for the input arguments? then maybe can throw the parseException if the formatter is not of the right pattern

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 12]
```diff
@@ -141,6 +144,28 @@ public static Set<Tag> parseTags(Collection<String> tags) throws ParseException
...
+        String trimmedDate = date.trim();
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
+        return LocalDate.parse(trimmedDate, formatter);
+    }
+
+    /**
+     * Parses {@code String time} into an {@code LocalTime} and returns it. Leading and trailing whitespaces will be
+     * trimmed.
+     */
+    public static LocalTime parseTime(String time) throws ParseException {
```
Comment (Respirayson) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:169]: here also I think should check if valid localtime string before trying to parse it as a LocalTime

File: src/main/java/seedu/address/model/Model.java [hunk 13]
```diff
@@ -171,6 +175,27 @@ public interface Model {
      */
     void updateFilteredPersonList(Predicate<Person> predicate);
 
+    /**
+     * Returns true if a task with the same identity as {@code task} exists in the task list.
```
Comment (Respirayson) [inline src/main/java/seedu/address/model/Model.java]: perhaps a typo in the javadoc

File: src/main/java/seedu/address/model/consultation/ConsultationList.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.Iterator;
+import java.util.List;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.consultation.exceptions.ConsultationNotFoundException;
+import seedu.address.model.consultation.exceptions.DuplicateConsultationException;
+
+/**
+ * A list of Consultation.
```
Comment (Respirayson) [inline src/main/java/seedu/address/model/consultation/ConsultationList.java]: Really small thing but its a list of Consultations (plural) oops

File: src/main/java/seedu/address/model/consultation/ReadOnlyConsultation.java [hunk 15]
```diff
@@ -0,0 +1,17 @@
+package seedu.address.model.consultation;
+
+import javafx.collections.ObservableList;
+import seedu.address.model.person.Person;
+
+/**
```
Comment (Respirayson) [inline src/main/java/seedu/address/model/consultation/ReadOnlyConsultation.java:6]: Javadoc here mismatch

File: src/main/java/seedu/address/storage/StorageManager.java [hunk 16]
```diff
@@ -23,18 +24,20 @@ public class StorageManager implements Storage {
     private UserPrefsStorage userPrefsStorage;
     private TaskListStorage taskListStorage;
     private SessionListStorage sessionListStorage;
+    private ConsultationListStorage consultationListStorage;
 
     /**
      * Creates a {@code StorageManager} with the given {@code AddressBookStorage}
```
Comment (Respirayson) [inline src/main/java/seedu/address/storage/StorageManager.java]: perhaps you could help to update the javadoc here

File: src/main/java/seedu/address/ui/ConsultationListPanel.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+     * Creates a {@code ConsultationListPanel} with the given {@code ObservableList}
+     */
+    public ConsultationListPanel(ObservableList<Consultation> consultationList) {
+        super(FXML);
+        consultationListView.setItems(consultationList);
+        consultationListView.setCellFactory(listView -> new ConsultationListViewCell());
+    }
+
+    /**
+     * Custom {@code ListCell} that displays the graphics of a {@code Task} using a {@code TaskCard}.
```
Comment (Respirayson) [inline src/main/java/seedu/address/ui/ConsultationListPanel.java]: mismatch in javadoc

File: src/test/java/seedu/address/logic/commands/AddCommandTest.java [hunk 18]
```diff
@@ -90,165 +79,6 @@ public void toStringMethod() {
         assertEquals(expected, addCommand.toString());
     }
 
-    /**
```
Comment (Respirayson) [inline src/test/java/seedu/address/logic/commands/AddCommandTest.java:93]: I like the extraction of the stub

File: src/test/java/seedu/address/model/person/PersonTest.java [hunk 19]
```diff
@@ -83,6 +76,18 @@ public void isSamePerson() {
         assertFalse(BOB.isSamePerson(editedBob));
     }
 
+    @Test
+    public void isSameName() {
```
Comment (Respirayson) [inline src/test/java/seedu/address/model/person/PersonTest.java:80]: I guess good to increase the scope. Good job

File: src/test/java/seedu/address/storage/JsonConsultationListStorageTest.java [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    @Test
+    public void saveConsultationListBook_nullFilePath_throwsNullPointerException() {
+        assertThrows(NullPointerException.class, () ->
+                saveConsultationListBook(new ConsultationListBook(), null));
+    }
+
+    /**
+     * Saves {@code addressBook} at the specified {@code filePath}.
```
Comment (Respirayson) [inline src/test/java/seedu/address/storage/JsonConsultationListStorageTest.java]: javadoc mismatch

**Review by Respirayson — COMMENTED** [review]
few nits

**Review by Respirayson — APPROVED** [review]
LGTM

**Pull request merged**